walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,159 @@
+    // List<TimeRange> availableTimeRanges = getAvailableTimeRanges(unavailableTimeRanges);
+
+    throw new UnsupportedOperationException("TODO: continue building schedule");
+  }
+
+  public List<Event> filterEvents(Collection<Event> events, MeetingRequest request) {
+    List<Event> filteredEvents = new ArrayList<Event>();
+    Collection<String> requiredAttendees = request.getAttendees();
+
+    //not the most efficient way, but it works
+    for (Event event: events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      for (String attendee: eventAttendees) {
+        if (requiredAttendees.contains(attendee)) {
+          filteredEvents.add(event);
+        }
+      }
+      
+    }
+    return filteredEvents;
+  }
+
+  public List<Event> getSortedTimeRanges(List<Event> events) {
+    // Collections.sort() only takes in List, so must convert events to a list
+    List<Event> eventsAsList= new ArrayList<Event>();
+    for (Event event: events) {
+      eventsAsList.add(event);
+    }
+    Collections.sort(eventsAsList, Event.ORDER_BY_START);
+    return eventsAsList;
+  }
+
+  public List<TimeRange> getUnavailableTimeRanges(List<Event> events) {
+    List<TimeRange> unavailableTimeRanges = new ArrayList<TimeRange>();
+    List<TimeRange> eventsAsTimeRanges = new ArrayList<TimeRange>();
+
+    for (Event event: events) {
+      eventsAsTimeRanges.add(event.getWhen());
+    }
+
+    //case 1: events that do not overlap
+    for (TimeRange timeRange_1: eventsAsTimeRanges) {
+      boolean timeRangesOverlap = false;
+      for (TimeRange timeRange_2: eventsAsTimeRanges)  {
+        if (!(timeRange_1.equals(timeRange_2))) {
+          if ((timeRange_1.overlaps(timeRange_2))) {
+            timeRangesOverlap = true;
+          }
+        }
+      }
+      if (timeRangesOverlap == false) {
+         // ensure no double-counting timeRanges
+        if (!(unavailableTimeRanges.contains(timeRange_1))){
+          unavailableTimeRanges.add(timeRange_1);
+        }
+      }
+    }
+
+    //very brute-force way to cover overlapping time ranges
+    for (TimeRange timeRange_1: eventsAsTimeRanges) {
+      for (TimeRange timeRange_2: eventsAsTimeRanges)  {
+        if (!(timeRange_1.equals(timeRange_2))) {
+          if ((timeRange_1.overlaps(timeRange_2))) {
+
+            // case 2: "regular" overlapping events
+            int timeR2Start = timeRange_2.start();
+            int timeR2End = timeRange_2.end();
+
+            //TODO: clean up the if statement below. basically check if 
+            //timeRange_1 starts before and ends during timeRange_2
+            if((timeRange_1.contains(timeR2Start)) && (!(timeRange_1.contains(timeR2End)))) {
+
+              TimeRange newTimeRange = TimeRange.fromStartEnd(timeRange_1.start(), timeR2End, false);
+              unavailableTimeRanges.add(newTimeRange);
+            }
+
+            // case 3: nested events
+            else if (timeRange_1.contains(timeRange_2)) {
```

Comment:
this line should be on line 149 like so:

`} else if (timeRange_1.contains(timeRange_2)) {`